index.js

```diff
@@ -48,96 +48,127 @@ function fromFile(file, callback) {
     });
 }
 
+
+function returnOutput(output) {
+    //check for tm2z
+    if (output.toString('ascii', 257, 262) === 'ustar') return 'tm2z';
+
+    //check for serial tiles
+    const head = output.slice(0, 50);
+    if (head.toString().indexOf('JSONBREAKFASTTIME') === 0) return 'serialtiles';
+
+    //check for tif+gz
+    if (
+        (output.slice(0, 2).toString() === 'II' || output.slice(0, 2).toString() === 'MM')
+        && ((output[2] === 42) || output[3] === 42 || output[2] === 43)) {
+        return 'tif+gz';
+    }
+
+    //default to unknown
+    throw invalid('Unknown filetype');
+}
+
+/**
+ * promisify zlib.gunzip.
+ *
+ * @param {buffer} buffer -
+ * @returns {promise}
+ */
+
+function gunzipPromise (buffer) {
+
+    if (buffer.Body) buffer = buffer.Body;
+
+    return new Promise((resolve, reject) => {
+        zlib.gunzip(buffer, {finishFlush: zlib.Z_SYNC_FLUSH }, (err, data) => {
+            if (err) return reject(err);
+            return resolve(data);
+        });
+    });
+};
+
 function detect(buffer, callback) {
-    var header = buffer.toString().substring(0, 400);
-
-    // check for topojson/geojson
-    if (header.trim().indexOf('{') == 0) {
-
-        // Remove spaces
-        var str = JSON.stringify(header);
-        var nospaces = str.replace(/\s/g, '');
-        header = JSON.parse(nospaces);
-
-        if (header.indexOf('\"tilejson\":') !== -1) return callback(null, 'tilejson');
-        if ((header.indexOf('\"arcs\":') !== -1) || (header.indexOf('\"objects\":') !== -1)) return callback(null, 'topojson');
-        if ((header.indexOf('\"features\":') !== -1) || (header.indexOf('\"geometries\":') !== -1) || (header.indexOf('\"coordinates\":') !== -1)) return callback(null, 'geojson');
-        if (header.indexOf('\"type\":') !== -1) {
-            var m = /"type":\s?"(.+?)"/.exec(header);
-            if (!m) {
-                return callback(invalid('Unknown filetype'));
+
+    gunzipPromise(buffer)
+      .then(data => {
+          return returnOutput(data);
+      })
+      .then(type => {
+        return callback(null, type);
+      })
+      .catch((err) => {
```

Comment:
@vincentsarago Wondering if this could be bad news in the case of a legitimate gunzip error. Mind adding a unit test that is a gunzipped file that is not valid or not geo data?